packages/next/src/server/app-render/csrf-protection.ts

```diff
@@ -0,0 +1,45 @@
+// micromatch is only available at node runtime, so it cannot be used here since the code path that calls this function
+// can be run from edge. This is a simple implementation that safely achieves the required functionality.
+// the goal is to match the functionality for remotePatterns as defined here -
+// https://nextjs.org/docs/app/api-reference/components/image#remotepatterns
+// TODO - retrofit micromatch to work in edge and use that instead
+function matchWildcardDomain(domain: string, pattern: string) {
+  const domainParts = domain.split('.')
+  const patternParts = pattern.split('.')
+
+  // Iterate through each part and compare them
+  for (let i = 0; i < patternParts.length; i++) {
+    if (patternParts[i] === '**') {
+      // If '**' is encountered, ensure remaining domain ends with remaining pattern
+      // e.g. **.y.z and a.b.c.y.z should match (b.c.y.z ends with y.z)
+      const remainingPattern = patternParts.slice(i + 1).join('.')
+      const remainingDomain = domainParts.slice(i + 1).join('.')
+      return remainingDomain.endsWith(remainingPattern)
+    } else if (patternParts[i] === '*') {
+      // If '*' is encountered, ensure remaining domain is equal to remaining pattern
+      // e.g. *.y.z and c.y.z should match (y.z is equal to y.z)
+      const remainingPattern = patternParts.slice(i + 1).join('.')
+      const remainingDomain = domainParts.slice(i + 1).join('.')
+      return remainingDomain === remainingPattern
+    }
+
+    // If '*' is not encountered, compare the parts
+    if (patternParts[i] !== domainParts[i]) {
+      return false
+    }
+  }
+
+  return true
+}
```

Comment:
```suggestion
  let patternComparisonString: string;
  
  if (patternParts.length < 1) {
    // pattern is empty and therefore invalid to match against
    return false
  } 
  
  if (domainParts.length < patternParts.length) {
    domain has too few segments and thus cannot match
    return false;
  }
  
  let depth = 0
  while (patternParts.length && depth++ < 2) {
    const pattern = patternParts.pop();
    const domain = domainParts.pop();
    
    switch (pattern) {
      case "":
      case "*":
      case "**" {
        // invalid pattern. pattern segments must be non empty
        // Additionally wildcards are only supported below the domain level
        return false
      }
      default: {
        if (domain !== pattern) {
          return false;
        }
      }
  }
  
  while (patternParts.length) {
    const pattern = patternParts.pop();
    const domain = domainParts.pop();
    
    switch (pattern) {
      case "": {
        // invalid pattern. pattern segments must be non empty
        return false
      }
      case "*": {
        // wildcard matches anything so we continue if the domain part is non-empty
        if (domain.length) {
          continue;
        } else {
          return false;
        }
      }
      case "**" {
        // if this is not the last item in the pattern the pattern is invalid
        if (patternParts.length > 0) {
          return false;
        }
        // recursive wildcard matches anything so we terminate here if the domain part is non empty
        return domain.length > 0
      }
      default: {
        if (domain !== pattern) {
          return false;
        }
      }
  }

  // We exhausted the pattern. If we also exhausted the domain we have a match
  return domainParts.length === 0
}
```

This implementation is a much lower overhead on memory allocations and iterations
It also is more strict, only allowing ** at the front and only supporting wildcards at the subdomain and deeper level